File: linked_list/0203_remove_linked_list_elements.js
/*
Given the head of a linked list and an integer val, remove all the nodes of the 
linked list that has Node.val == val, and return the new head.
*/

// time complexity O(n)
// space complexity O(1)
var removeElements = function (head, val) {
  let dummyNode = new ListNode(0, head);
  let curr = dummyNode;
  while (curr.next) {
    if (curr.next.val == val) {
      curr.next = curr.next.next;
    } else {
      curr = curr.next;
    }
  }
  return dummyNode.next;
};

// recursion
// time complexity O(n)
// space compleixty/calling stack O(n)
var removeElements = function (head, val) {
  if (head === null) {
    return head;
  }
  head.next = removeElements(head.next, val);
  return head.val === val ? head.next : head;
};

File: linked_list/remove_elements/0203_remove_linked_list_elements.py
# Given the head of a linked list and an integer val, remove all the nodes of the 
# linked list that has Node.val == val, and return the new head.

# https://leetcode.com/problems/remove-linked-list-elements/

class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy_head = ListNode(0, head)
        curr = dummy_head
        while curr.next:
            if curr.next.val == val:
                curr.next = curr.next.next
            else:
                curr = curr.next
        return dummy_head.next

# two pointers    
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy_head = ListNode(0, head)
        slow, fast = dummy_head, head
        while fast:
            # move fast to the next non-val node
            while fast and fast.val == val:
                fast = fast.next
            # if skip any nodes, remove them 
            if slow.next is not fast:
                slow.next = fast
            # move slow
            slow = slow.next
            # move fast
            fast = fast.next if fast else None
        return dummy_head.next
    
# recursion
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        # base case
        if not head:
            return None
        # recursive case
        head.next = self.removeElements(head.next, val)
        # remove current node or not
        if head.val == val:
            # remove current node
            return head.next   
        else:
            # keep current node
            return head